File: editorial/chap04/answer_B19.cpp
#include <iostream>
#include <algorithm>
using namespace std;

long long N, W, w[109], v[109];
long long dp[109][100009];

int main() {
	// 入力・配列の初期化
	cin >> N >> W;
	for (int i = 1; i <= N; i++) cin >> w[i] >> v[i];
	for (int i = 0; i <= N; i++) {
		for (int j = 0; j <= 100000; j++) dp[i][j] = 1'000'000'000'000'000LL;
	}

	// 動的計画法
	dp[0][0] = 0;
	for (int i = 1; i <= N; i++) {
		for (int j = 0; j <= 100000; j++) {
			if (j < v[i]) dp[i][j] = dp[i - 1][j];
			else dp[i][j] = min(dp[i - 1][j], dp[i - 1][j - v[i]] + w[i]);
		}
	}

	// 答えの出力
	long long Answer = 0;
	for (int i = 0; i <= 100000; i++) {
		if (dp[N][i] <= W) Answer = i;
	}
	cout << Answer << endl;
	return 0;
}

File: editorial/chap04/python/answer_B19.py
# 入力
N, W = map(int, input().split())
w = [ None ] * (N + 1)
v = [ None ] * (N + 1)
for i in range(1, N+1):
	w[i], v[i] = map(int, input().split())

# 動的計画法
dp = [ [ 10 ** 15 ] * 100001 for i in range(N + 1) ]
dp[0][0] = 0
for i in range(1, N+1):
	for j in range(100001):
		if j < v[i]:
			dp[i][j] = dp[i-1][j]
		if j >= v[i]:
			dp[i][j] = min(dp[i-1][j], dp[i-1][j-v[i]]+w[i])

# 出力
Answer = 0
for i in range(100001):
	if dp[N][i] <= W:
		Answer = i
print(Answer)
